rust/agama-server/src/scripts/web.rs

```diff
@@ -0,0 +1,139 @@
+    scripts::{Script, ScriptError, ScriptsGroup, ScriptsRepository},
+};
+use axum::{
+    extract::State,
+    http::StatusCode,
+    response::{IntoResponse, Response},
+    routing::{get, post},
+    Json, Router,
+};
+use serde::{Deserialize, Serialize};
+use serde_json::json;
+use thiserror::Error;
+use tokio::sync::RwLock;
+
+#[derive(Clone, Default)]
+struct ScriptsState {
+    scripts: Arc<RwLock<ScriptsRepository>>,
+}
+
+#[derive(Error, Debug)]
+#[error("Script error: {0}")]
+struct ScriptServiceError(#[from] ScriptError);
+
+impl IntoResponse for ScriptServiceError {
+    fn into_response(self) -> Response {
+        let body = json!({
+            "error": self.to_string()
+        });
+        (StatusCode::BAD_REQUEST, Json(body)).into_response()
+    }
+}
+
+/// Sets up and returns the axum service for the auto-installation scripts.
+pub async fn scripts_service() -> Result<Router, ServiceError> {
+    let state = ScriptsState::default();
+    let router = Router::new()
+        .route(
+            "/",
+            get(list_scripts).post(add_script).delete(remove_scripts),
+        )
+        .route("/run", post(run_scripts))
+        .with_state(state);
+    Ok(router)
+}
+
+#[utoipa::path(
+    post,
+    path = "/",
+    context_path = "/api/scripts",
+    request_body(content = ScriptConfig, description = "Script definition"),
+    responses(
+        (status = 200, description = "The script was added.")
+    )
+)]
+async fn add_script(
+    state: State<ScriptsState>,
+    Json(script): Json<Script>,
+) -> Result<impl IntoResponse, ScriptServiceError> {
+    let mut scripts = state.scripts.write().await;
+    scripts.add(script);
+    Ok(())
+}
+
+#[utoipa::path(
+    get,
+    path = "/",
+    context_path = "/api/scripts",
+    responses(
+        (status = 200, description = "Defined scripts.")
+    )
+)]
+async fn list_scripts(state: State<ScriptsState>) -> Json<Vec<Script>> {
+    let repo = state.scripts.read().await;
+    Json(repo.scripts.to_vec())
+}
+
+#[utoipa::path(
+    delete,
+    path = "/",
+    context_path = "/api/scripts",
+    responses(
+        (status = 200, description = "The scripts have been removed.")
+    )
+)]
+async fn remove_scripts(
+    state: State<ScriptsState>,
+) -> Result<impl IntoResponse, ScriptServiceError> {
+    let mut scripts = state.scripts.write().await;
+    scripts.clear();
+    Ok(())
+}
+
+#[derive(Clone, Serialize, Deserialize)]
+struct RunScriptParams {
+    group: ScriptsGroup,
+}
+
+#[utoipa::path(
+    post,
+    path = "/run",
+    context_path = "/api/scripts",
+    responses(
+        (status = 200, description = "The scripts were successfully executed.")
+    )
+)]
+async fn run_scripts(
+    state: State<ScriptsState>,
+    Json(group): Json<ScriptsGroup>,
+) -> Result<(), ScriptServiceError> {
+    let scripts = state.scripts.write().await;
+    if let Err(error) = scripts.run(group).await {
+        tracing::error!("Could not run user-defined scripts: {error}");
```

Comment:
so it means that it will return 200 when coulld not run scripts? I do not feel it is good behavior.

Reply:
I think it is OK if one of the scripts fails. You run, log the error, and continue with the next one.

Reply:
Um, I'm not sure. Maybe some script could depend on the result of the previous one.

Maybe it should be configurable whether to continue on error or not. :thinking: